### ISSUE TITLE
[BUG][Kotlin] Generated clients cannot serialize into `List<T>`

### ISSUE THREAD

**Issue by trietsch** [OpenAPITools/openapi-generator#8560]
##### Description

An endpoint definition in the Open Api Spec that returns an array of objects, is not correctly serialized by all available `serializationLibrary` options (Moshi, Gson, Jackson) for the Kotlin generator.
This results in deserialization issues, such as:
```
Exception in thread "main" java.lang.ClassCastException: class com.squareup.moshi.LinkedHashTreeMap cannot be cast to class com.some.package.Example (com.squareup.moshi.LinkedHashTreeMap and com.some.package.Example are in unnamed module of loader 'app')
```

##### openapi-generator version

Version 5.0.0

##### OpenAPI declaration file content or url

```yaml
openapi: "3.0.0"
info:
  version: 0.1.0
  title: Example List<T> bug

paths:
  "/admin/streams":
    get:
      summary: "Get all examples"
      responses:
        200:
          description: "The examples"
          content:
            application/json:
              schema:
                type: array
                items:
                  $ref: "#/components/schemas/Example"

components:
  schemas:
    Example:
      type: object
      properties:
        name:
          type: string
        amount:
          type: number
```

##### Generation Details

For generating, the only thing that is important, is using the `Kotlin` generator. Other options do not have any effect on the bug, as it's unrelated. Regardless of the `serializationLibrary`, the issue persists over all of them.

##### Steps to reproduce

Use the generated client to perform a call that returns a list of objects. The call succeeds, however, if you'd want to do (e.g.) a `map` operation, that will fail, as Kotlin tries to cast the object there.

##### Related issues/PRs

None

##### Suggest a fix

I'll create a Pull Request soon, with the fixes for all `serializationLibrary` options. For Jackson, [I've already created a fix](https://github.com/streammachineio/openapi-generator/commit/f7adf0974e7f774909e69a31bfd1a7474f830e4d) (which was by using a `TypeReference<T>`, instead of `T::class.java`.

**Comment by adamsz-lume** [comment]
Hi,
How is the fixing going? I have just hit this bug and it seems to be quite a big one considering it makes the client unusable.

I can chip in with the fix maybe. Do you have a branch I can have a look at?

**Comment by adamsz-lume** [comment]
I have googled this solution. You might want to have a look.
```kotlin
val mainType = typeOf<T>()
                    val finalType = Types.newParameterizedType(
                        T::class.java,
                        *mainType.arguments.map { it.type!!.javaType }.toTypedArray()
                    )
                    return Serializer.moshi.adapter<T>(finalType).fromJson(bodyContent)
```
I have raised that in moshi's repo so maybe they will introduce a convenient function for it.

**Comment by trietsch** [comment]
Hey @adamsz-lume! Indeed, this bug makes pretty much all clients unusable. 
It would be great if you could take a look. Did you see the PR at https://github.com/OpenAPITools/openapi-generator/pull/8796 ?

The branch van be found in the form I made: https://github.com/streammachineio/openapi-generator/tree/add-support-for-collection-of-generic-classes

If you'd want to contribute to this fix, I propose you fork the form that I made, and make a PR to my branch. Next, I'll update the open PR for the OpenApiGenerator project. 

Let me know if and when you'll pick this up, then I'll mark my PR for the OpenApiGenerator as draft, so that it won't be merged yet.

**Comment by trietsch** [comment]
@adamsz-lume how's the fix for Moshi going?

**Comment by adamsz-lume** [comment]
the PR has been waiting for you https://github.com/streammachineio/openapi-generator/pull/1 :)

**Comment by Mordag** [comment]
What is the status of this PR? Would be awesome to have this part of the next release.

**Comment by trietsch** [comment]
The PR is ready for quite some time now. I added support for Jackson and Gson, and @adamsz-lume added support for Moshi. Both changes are included in this PR. 

Only task that remains is a review by someone from the core team.

**Comment by Mordag** [comment]
I think you should have a look at why the CI is failing first. Usually it is something minor you need to change to unblock that part (try doing a rebase - just in case). Then tag the relevant team members again and make clear that the PR is ready for review/merge.

**Comment by trietsch** [comment]
Ah, it wasn't failing first. I don't know exactly when I have time to check it, probably somewhere this week.

**Comment by Mordag** [comment]
Yep, for example:

```
//change the request type to string - you will need to change the ApiClient.responseBody with `if (T::class == String::class)` method to instantly return the raw object as string if you requested a return type of string.
val localVarResponse = request<String>(
            localVariableConfig
        )

        return when (localVarResponse.responseType) {
            ResponseType.Success -> {
                //Add this part
                val rawString = (localVarResponse as Success<*>).data as String
                val listType: Type = Types.newParameterizedType(List::class.java, MyObject::class.java)
                val adapter: JsonAdapter<List<MyObject>> = Serializer.moshi.adapter(listType)
                adapter.fromJson(rawString) as List<MyObject>
            }
            ...
        }
```

**Comment by missionlanejeremymcbride** [comment]
Any updates on when the PR will be merged for this fix?

**Comment by trietsch** [comment]
Thanks for the reminder, I'll try to pick it up today or tomorrow.

**Comment by missionlanejeremymcbride** [comment]
Thanks, @trietsch!

**Comment by trietsch** [comment]
@missionlanejeremymcbride new PR is ready for review by the OpenApi team: https://github.com/OpenAPITools/openapi-generator/pull/9918

**Comment by ehuang5837** [comment]
I've been keep an eye on this issue and saw recently the PR was merged, after upgrading the openapi generator version i'm running into this issue
```Platform class kotlin.Unit requires explicit JsonAdapter to be registered```
seems to be some issue with the moshi adapter, does anyone happen to know a quick fix?

**Comment by 4brunu** [comment]
Maybe @adamsz-lume can help with the error?

**Comment by peakay** [comment]
having the same problem with moshi @ehuang5837, swapped to using gson serialization and it was working


see serializationLibrary option https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/kotlin.md
